src/commands/flutter-symbols/upload.ts

```diff
@@ -0,0 +1,468 @@
+        callResults.push(await this.performDsymUpload())
+      }
+      if (this.androidMappingLocation) {
+        callResults.push(await this.performAndroidMappingUpload(metricsLogger, apiKeyValidator))
+      }
+      if (this.dartSymbolsLocation) {
+        callResults.push(...(await this.performDartSymbolsUpload(metricsLogger, apiKeyValidator)))
+      }
+
+      const totalTime = (Date.now() - initialTime) / 1000
+      this.context.stdout.write(renderCommandSummary(callResults, totalTime, this.dryRun))
+    } catch (e) {
+      this.context.stderr.write(renderGeneralizedError(e))
+
+      return 1
+    } finally {
+      try {
+        await metricsLogger.flush()
+      } catch (err) {
+        this.context.stdout.write(`WARN: ${err}\n`)
+      }
+    }
+
+    return 0
+  }
+
+  private getAndroidMetadata(): MappingMetadata {
+    return this.getMappingMetadata(TYPE_JVM_MAPPING)
+  }
+
+  private getFlutterMetadata(platform: string, arch: string) {
+    return this.getMappingMetadata(TYPE_DART_SYMBOLS, platform, arch)
+  }
+
+  private getFlutterSymbolFiles(dartSymbolLocation: string): string[] {
+    const symbolPaths = glob.sync(buildPath(dartSymbolLocation, '*.symbols'))
+
+    return symbolPaths
+  }
+
+  private getGitDataPayload(): MultipartValue | undefined {
+    if (this.gitData === undefined) {
+      return undefined
+    }
+
+    const files = this.gitData.trackedFilesMatcher.rawTrackedFilesList()
+    const repoPayload = {
+      data: [
+        {
+          files,
+          hash: this.gitData.hash,
+          repository_url: this.gitData.remote,
+        },
+      ],
+      version: 1,
+    }
+
+    return {
+      options: {filename: 'repository', contentType: 'application/json'},
+      value: JSON.stringify(repoPayload),
+    }
+  }
+
+  private async getGitMetadata(): Promise<RepositoryData | undefined> {
+    try {
+      return await getRepositoryData(await newSimpleGit(), this.repositoryUrl)
+    } catch (e) {
+      this.context.stdout.write(renderGitWarning(e))
+    }
+
+    return undefined
+  }
+
+  private getMappingMetadata(type: string, platform?: string, arch?: string): MappingMetadata {
+    return {
+      arch,
+      cli_version: this.cliVersion,
+      git_commit_sha: this.gitData?.hash,
+      git_repository_url: this.gitData?.remote,
+      platform,
+      service: this.serviceName,
+      type,
+      variant: this.flavor,
+      version: this.getSanitizedVersion(),
+    }
+  }
+
+  private getSanitizedVersion() {
+    return this.version!.replace('+', '-')
+  }
+
+  private async parsePubspecVersion(pubspecLocation: string): Promise<number> {
+    if (!fs.existsSync(pubspecLocation)) {
+      this.context.stderr.write(renderMissingPubspecError(pubspecLocation))
+
+      return 1
+    }
+
+    try {
+      const doc = yaml.load(fs.readFileSync(pubspecLocation, 'utf8')) as any
+      if (doc.version) {
+        this.version = doc.version
+
+        const buildIndex = this.version!.indexOf('+')
```

Comment:
The parsing logic of semantic versions may have edge cases that are not obvious to cover.

We may want to add the `semver` package to parse the version in a more robust fashion? (it is already present in the node_modules as it is imported by `dd-trace` at least, so it should not increase the size of the project)